src/Cello/problem_MethodFluxCorrect.cpp

```diff
@@ -203,120 +227,290 @@ void MethodFluxCorrect::compute_continue_sum_fields
+
+    // load the density array
+    cello_float* density_array = nullptr;
+    if (field.is_field("density")){
+      density_array = (cello_float*) field.unknowns("density");
+
+      // copy the values in the density_array (we could be more selective about
+      // what we copy)
+      for (int iz=0; iz<nz; iz++) {
+        for (int iy=0; iy<ny; iy++) {
+          for (int ix=0; ix<nx; ix++) {
+            int i=ix + mx*(iy + my*iz);
+            old_rho_array[i] = density_array[i];
           }
         }
       }
-      
-      if (rank >= 2) {
-        axis=1;
-        // Y axis
-        for (face=0; face<2; face++) {
-          level_face = block->face_level(axis,face);
-          if (level_face > level) {
-            int dbx,dby,dbz;
-            int dnx,dny,dnz;
-            auto block_fluxes    = flux_data->block_fluxes(axis,face,i_f);
-            auto neighbor_fluxes = flux_data->neighbor_fluxes(axis,face,i_f);
-            auto block_flux_array =
-              block_fluxes->flux_array(&dbx,&dby,&dbz);
-            auto neighbor_flux_array =
-              neighbor_fluxes->flux_array(&dnx,&dny,&dnz);
-          
-            iy = (face == 0) ? 0 : ny-1;
-            for (iz=0; iz<nz; iz++) {
-              for (ix=0; ix<nx; ix++) {
-                int i=ix+mx*(iy+my*iz);
-                int ib = ix*dbx+iz*dbz;
-                int in = ix*dnx+iz*dnz;
-                field_array[i] +=
-                  (2*face-1)*(block_flux_array[ib] - neighbor_flux_array[in]);
-              }
+    }
+
+    // perform the density flux correction
+    if (i_f_density > -1){
+      flux_correct_helper_(density_array, mx,my,mz, nx,ny,nz,
+                           i_f_density, flux_data, rank, perform_correction);
+    }
+
+
+    // perform the flux corrections for the other fields
+    Grouping * groups = cello::field_groups();
+    for (int i_f=0; i_f<nf; i_f++) {
+      const int index_field = flux_data->index_field(i_f);
+      const std::string field_name = field.field_name(index_field);
+
+      if (i_f == i_f_density){ // density flux correction already happened
+        continue;
+      }
+
+      cello_float* field_array = (cello_float*) field.unknowns(index_field);
+
+      if (groups->is_in(field_name, "make_field_conservative")){
+        // Handle flux corrections for fields that must be multiplied by the
+        // density to be made conservative
+
+        ASSERT1("MethodFluxCorrect::flux_correct_",
+                ("The \"density\" field must exist to perform flux "
+                 "corrections on \"%s\"."), field_name.c_str(),
+                density_array != nullptr);
+
+        // compute the conserved quantity
+        for (int iz=0; iz<nz; iz++) {
+          for (int iy=0; iy<ny; iy++) {
+            for (int ix=0; ix<nx; ix++) {
+              int i=ix + mx*(iy + my*iz);
+              temp_cons_array[i] = old_rho_array[i]*field_array[i];
```

Comment:
Exactly! In other words, to apply flux corrections for cells adjacent to the x-face:

- ρ_new = ρ_old + (dt/dx)*δF_x
- v_x,new = (ρ_old*v_x,old + (dt/dx)*δF_x)/ρ_new

Note: In Enzo-E, the factor of (dt/dx) is absorbed into the coarse and fine fluxes (I think this might help when you have adaptive time stepping in a cosmological simulation)